add wlm feature overview (#8632)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/workload-management/workload-management.md

---
layout: default
title: Workload management
nav_order: 62
has_children: true
parent: Availability and recovery
---

# Workload management

Workload management allows users to group and search network traffic, isolating system resources to prevent the overuse of network resources by specific requests. It offers the following benefits:
 
- Tenant-level admission control and reactive query management. It can identify and cancel resource-intensive queries when configured thresholds are exceeded, ensuring fair resource allocation.

- Tenant-level isolation within the cluster for search workloads, operating at a node level.

System administrators can dynamically manage query groups using the Workload management APIs. These query groups can be used to make search requests with resource limits.

### QueryGroup

```json

> {

Reviewer: Not sure what this example does. Does it create a query group? If so, should there be an API operation attached to this?

Reviewer: I replaced this with an example from the Query Groups page.

Reviewer 2: I believe the original version is to show the schema of a query group instead of creating one

| `wlm.query_group.enforcement_interval`  | Defines the monitoring interval. |
| `wlm.query_group.mode`  | Defines the feature operating mode. |
| `wlm.query_group.node.memory_rejection_threshold` | Defines the query group level `memory` threshold. When the threshold is exceeded, the request is rejected.  |
| `wlm.query_group.node.cpu_rejection_threshold` | Defines query group level `cpu` threshold.  When the threshold is exceeded, the request is rejects. |
| `wlm.query_group.node.memory_cancellation_threshold` | Controls whether the node is considered in duress when the `cpu` threshold is reached and the effective request cancellation threshold based on `memory` usage. |
| `wlm.query_group.node.cpu_cancellation_threshold`    | Controls whether the node is considered in duress when the `cpu` threshold is reached and the effective request cancellation threshold on `cpu` usage. |
 
When setting rejection andcancellation settings thresholds, remember that the rejection threshold for a resource should always be less than the cancellation threshold. 

### Workload management stats

> The stats API is useful to gather current workload management metrices at query group level. The stats API looks like following

Reviewer: Workload stats API?

Author: these stats are specific to this feature, workload stats is a very broad term and implicitly refers to indexing and search workload. Here we just mean the tenant level search workload stats.

> ```json

Reviewer: Should this be on the API page instead of the overview page?

```json
GET _wlm/stats
```

## Example response body

```json
{
  "_nodes": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "cluster_name": "XXXXXXYYYYYYYY",
  "A3L9EfBIQf2anrrUhh_goA": {
    "query_groups": {
      "16YGxFlPRdqIO7K4EACJlw": {
        "total_completions": 33570,
        "total_rejections": 0,
        "total_cancellations": 0,
        "cpu": {
          "current_usage": 0.03319935314357281,
          "cancellations": 0,
          "rejections": 0
        },
        "memory": {
          "current_usage": 0.002306486276211217,
          "cancellations": 0,
          "rejections": 0
        }
      },
      "DEFAULT_QUERY_GROUP": {
        "total_completions": 42572,
        "total_rejections": 0,
        "total_cancellations": 0,
        "cpu": {
          "current_usage": 0,
          "cancellations": 0,
          "rejections": 0
        },
        "memory": {
          "current_usage": 0,
          "cancellations": 0,
          "rejections": 0
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

### Response body fields 
| Field name | Description |
|:----|:--- | 

> | `total_completions`   | The total number of request completions in this `query_group` at the given node. This includes all shard-level and co-ordinator-level requests. |

Reviewer: ```suggestion
| `total_completions`   | The total number of request completions in this `query_group` at the given node. This includes all shard-level and coordinator-level requests. |
```

> | `total_rejections`    | The total number request rejections in this `query_group` at the given node. This includes all shard-level and co-ordinator-level requests.    |

Reviewer: ```suggestion
| `total_rejections`    | The total number request rejections in this `query_group` at the given node. This includes all shard-level and coordinator-level requests.    |
```

> | `total_cancellations` | The total number of cancellations in this `query_group` at the given node.  This includes all shard-level and co-ordinator-level requests.  |

Reviewer: ```suggestion
| `total_cancellations` | The total number of cancellations in this `query_group` at the given node.  This includes all shard-level and coordinator-level requests.  |
```

---
layout: default
title: Workload management
nav_order: 70
has_children: true
parent: Availability and recovery
---

# Workload management

Workload management allows users to group and search network traffic, isolating system resources to prevent the overuse of network resources by specific requests. It offers the following benefits:
 
- Tenant-level admission control and reactive query management. It can identify and cancel resource-intensive queries when the configured thresholds are exceeded, ensuring fair resource allocation.

- Tenant-level isolation within the cluster for search workloads, operating at a node level.

## Query groups

System administrators can dynamically manage query groups using the Workload management APIs. These query groups can be used to make search requests with resource limits.

> The following example adds a query group with the named `analytics`:

Reviewer: ```suggestion
Query groups use the following schema:
```

```json

> ```json

Reviewer: ```suggestion
```

> PUT _wlm/query_group

Reviewer: ```suggestion
```

        "total_rejections": 0,
        "total_cancellations": 0,
        "cpu": {
          "current_usage": 0.03319935314357281,
          "cancellations": 0,
          "rejections": 0
        },
        "memory": {
          "current_usage": 0.002306486276211217,
          "cancellations": 0,
          "rejections": 0
        }
      },
      "DEFAULT_QUERY_GROUP": {
        "total_completions": 42572,
        "total_rejections": 0,
        "total_cancellations": 0,
        "cpu": {
          "current_usage": 0,
          "cancellations": 0,
          "rejections": 0
        },
        "memory": {
          "current_usage": 0,
          "cancellations": 0,
          "rejections": 0
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

### Response body fields 
| Field name | Description |
|:----|:--- | 
| `total_completions`   | The total number of request completions in this `query_group` at the given node. This includes all shard-level and coordinator-level requests. |
| `total_rejections`    | The total number request rejections in this `query_group` at the given node. This includes all shard-level and coordinator-level requests.    |

> | `total_cancellations` | The total number of cancellations in this `query_group` at the given node.  This includes all shard-level and coordinator-level requests.  |

Reviewer: ```suggestion
| `total_cancellations` | The total number of cancellations in this `query_group` at the given node. This includes all shard-level and coordinator-level requests.  |
```
